fqd_ccs.c

```diff
@@ -240,24 +239,26 @@ fqd_ccs_loop(remote_client *client) {
     if(rv > 0) poll_timeout = 10;
     else poll_timeout *= 2;
     if(poll_timeout > 4000) poll_timeout = 4000;
+    /* we must wake up often enough to emit our heartbeat */
+    if(client->heartbeat_ms && poll_timeout > client->heartbeat_ms) poll_timeout = client->heartbeat_ms;
     t = fq_gethrtime();
-    hb_us = ((unsigned long long)client->heartbeat_ms) * 1000000ULL;
-    if(client->heartbeat_ms &&
-       (unsigned long long)client->last_heartbeat < (unsigned long long)(t - hb_us)) {
-      if(fqd_css_heartbeat(client)) break;
+    unsigned long long hb_ns = ((unsigned long long)client->heartbeat_ms) * 1000000ULL;
+    long long hb_age_ns = t - client->last_heartbeat;
+    if(client->heartbeat_ms && hb_age_ns > hb_ns) {
+      if(fqd_ccs_heartbeat(client)) break;
       client->last_heartbeat = t;
     }
-    if(hb_us &&
-       (unsigned long long)client->last_activity < (unsigned long long)(t - hb_us * 3)) {
+    if(hb_ns && client->last_activity < (t - hb_ns * 3)) {
```

Comment:
`last_activity` isn't updated until below and the heartbeat does not touch `last_activity`.. is this error message misleading?

Reply:
We're using last_heartbeat for *sending* heartbeats, whereas last_activity is used for tracking inbound messages (including heartbeats).... so last_activity indicates that we've not seen any inbound messages in "too long".  I could change `last_heartbeat` to `last_outbound_heartbeat`; would that make it clearer?

Reply:
No, I misunderstood the error message.. We are telling the client we haven't *received* a HB from them in some timeout.  Maybe:

```
 ERRTOFD(client->fd, "heartbeat recv failed");
```
?